### ISSUE TITLE
Failed promise before calling allOf is not failing the wrapped promise.

### ISSUE THREAD

**Issue by awx-michael-wang** [temporalio/sdk-java#2046]
## Expected Behavior
`Promise.allOf(promises)` should fail as long as one of the `promises` fails, no matter it is already failed before calling the `allOf` or after.

## Actual Behavior
If one `Promise` in `promises` is already failed before calling `allOf`, the failure is ignored.

## Steps to Reproduce the Problem

```kotlin
val promise1 = Async.procedure { throw RuntimeException() }
val promise2 = Async.procedure { // do something else }

Promise.allOf(promise1, promise2).get()
```

## Specifications

  - Version: 1.20.0

**Comment by Quinn-With-Two-Ns** [comment]
Yes this is a bug, looks like we ignore the promise if it has completed regardless of how it completed

https://github.com/temporalio/sdk-java/blob/ed211fa611112288b576a2c979be9284e17fec89/temporal-sdk/src/main/java/io/temporal/internal/sync/AllOfPromise.java#L54 

We should check if it has completed exceptionally or not.

Fixing this though could break backwards compatibility since this could effect workflow code.  Any fix will need to make sure to preserve history compatibility with workflows that have this scenario.

**Comment by Quinn-With-Two-Ns** [comment]
In the mean time the easiest way to workaround this bug I think is to implement a wrapper around `promiseAllOf` to check if any promise passed in failed 

``` java
static Promise<Void> allOf(Promise<?>... promises) {
  for (Promise<?> p : promises) {
      if (p.isCompleted() && p.getFailure() != null) {
        return Workflow.newFailedPromise(p.getFailure());
      }
    }
    return WorkflowInternal.promiseAllOf(promises);
  }
}
```

**Comment by awx-michael-wang** [comment]
Thank you. I did implement a similar wrapper:
```
fun <T> List<Promise<T>>.safeAllOf(): Promise<Void> {
    val existFailure = this.filter { it.isCompleted }
        .firstNotNullOfOrNull { it.failure }

    return if (existFailure != null) {
        Workflow.newFailedPromise(existFailure)
    } else {
        Promise.allOf(this)
    }
}
```
